server/asset/dcr/script.go

```diff
@@ -0,0 +1,328 @@
+
+const (
+	scriptP2PKH = 1 << iota
+	scriptP2SH
+	scriptStake
+	scriptMultiSig
+	scriptUnsupported
+)
+
+const pubkeyLength = 33 // Length of a serialized compressed pubkey.
+
+func (s dcrScriptType) isP2SH() bool {
+	return s&scriptP2SH != 0
+}
+
+func (s dcrScriptType) isP2PKH() bool {
+	return s&scriptP2PKH != 0
+}
+
+func (s dcrScriptType) isStake() bool {
+	return s&scriptStake != 0
+}
+
+// func (s dcrScriptType) isMultiSig() bool {
+// 	return s&scriptMultiSig != 0
+// }
+
+const (
+	// SwapContractSize is the worst case scenario size for a swap contract,
+	// which is the pk-script of the non-change output of an initialization
+	// transaction as used in execution of an atomic swap.
+	// See extractSwapAddresses for a breakdown of the bytes.
+	SwapContractSize = 97
+	// P2PKHSigScriptSize is the worst case (largest) serialize size
+	// of a transaction input script that redeems a compressed P2PKH output.
+	// It is calculated as:
+	//
+	//   - OP_DATA_73
+	//   - 72 bytes DER signature + 1 byte sighash
+	//   - OP_DATA_33
+	//   - 33 bytes serialized compressed pubkey
+	P2PKHSigScriptSize = 1 + 73 + 1 + 33
+	// All pubkey scripts are assumed to be version 0.
+	currentScriptVersion = 0
+)
+
+// parseScriptType creates a dcrScriptType bitmap for the script type. A script
+// type will be some combination of pay-to-pubkey-hash, pay-to-script-hash,
+// and stake. If a script type is P2SH, it may or may not be mutli-sig.
+func parseScriptType(scriptVersion uint16, pkScript, redeemScript []byte) dcrScriptType {
+	if scriptVersion != 0 {
+		return scriptUnsupported
+	}
+	var scriptType dcrScriptType
+	switch {
+	case isPubKeyHashScript(pkScript):
+		scriptType |= scriptP2PKH
+	case isScriptHashScript(pkScript):
+		scriptType |= scriptP2SH
+	case isStakePubkeyHashScript(pkScript):
+		scriptType |= scriptP2PKH | scriptStake
+	case isStakeScriptHashScript(pkScript):
+		scriptType |= scriptP2SH | scriptStake
+	default:
+		return scriptUnsupported
+	}
+	if scriptType.isP2SH() && txscript.IsMultisigScript(redeemScript) {
+		scriptType |= scriptMultiSig
+	}
+	return scriptType
+}
+
+type dcrScriptAddrs struct {
+	pubkeys   []dcrutil.Address
+	numPK     int
+	pkHashes  []dcrutil.Address
+	numPKH    int
+	nRequired int
+}
+
+// Extract the addresses from the pubkey script, or the redeem script if the
+// pubkey script is P2SH. Addresses can be of several types, but the types
+// suppported will be pubkey
+func extractScriptAddrs(scriptType dcrScriptType, pkScript, redeemScript []byte) (*dcrScriptAddrs, error) {
+	var err error
+	var addrs []dcrutil.Address
+	var numRequired int
+	pubkeys := make([]dcrutil.Address, 0)
+	pkHashes := make([]dcrutil.Address, 0)
+	// For P2SH and non-P2SH multi-sig, pull the addresses from the pubkey script.
+	if scriptType.isP2PKH() {
+		_, addrs, numRequired, err = txscript.ExtractPkScriptAddrs(0, pkScript, chainParams)
+	} else {
+		_, addrs, numRequired, err = txscript.ExtractPkScriptAddrs(0, redeemScript, chainParams)
```

Comment:
Sorry to harp on this, but I think this function could look like this:

```go
// Extract the addresses from the pubkey script, or the redeem script if the
// pubkey script is P2SH. Addresses can be of several types, but the types
// suppported will be pubkey.
func extractScriptAddrs(script []byte) (*dcrScriptAddrs, error) {
	// For P2SH and non-P2SH multi-sig, pull the addresses from the pubkey
	// script.
	_, addrs, numRequired, err := txscript.ExtractPkScriptAddrs(0, script, chainParams)
	if err != nil {
		return nil, fmt.Errorf("extractScriptAddrs: %v", err)
	}

	var pubkeys, pkHashes []dcrutil.Address
	for _, addr := range addrs {
		// If the address is an unhashed public key, is won't need a pubkey as
		// part of its sigScript, so count them separately.
		_, isPubkey := addr.(*dcrutil.AddressSecpPubKey)
		if isPubkey {
			pubkeys = append(pubkeys, addr)
		} else {
			pkHashes = append(pkHashes, addr)
		}
	}

	return &dcrScriptAddrs{
		pubkeys:   pubkeys,
		numPK:     len(pubkeys),
		pkHashes:  pkHashes,
		numPKH:    len(pkHashes),
		nRequired: numRequired,
	}, nil
}
```

The caller knows what script they want addresses extracted from.

Reply:
Actually, I think the if-else on the type assertion should perhaps be a type switch with a default.

```go
switch addr.(type) {
case *dcrutil.AddressSecpPubKey:
	pubkeys = append(pubkeys, addr)
case *dcrutil.AddressPubKeyHash:
	pkHashes = append(pkHashes, addr)
default: // *dcrutil.AddressSecSchnorrPubKey, *dcrutil.AddressScriptHash, etc.
	others = append(others, adddr)
}
```

Reply:
I moved the script selection to the caller. 

We're really only interested in figuring out if it's a straight pubkey or a pubkey hash to calculate script length and verify addresses. Since all address types except `AddressSecpPubKey` use `Hash160` to hash the pubkeys, there's no real need further classification.